pkg/fileutils/fileutils.go

```diff
@@ -1,33 +1,77 @@
 package fileutils
 
 import (
+	"errors"
 	"fmt"
 	"io"
 	"io/ioutil"
 	"os"
 	"path/filepath"
+	"strings"
 
 	"github.com/Sirupsen/logrus"
 )
 
 // Matches returns true if relFilePath matches any of the patterns
+// and isn't excluded by any of the subsequent patterns.
+func exclusion(pattern string) bool {
+	return pattern[0] == '!'
+}
+
+func empty(pattern string) bool {
+	return pattern == ""
+}
+
 func Matches(relFilePath string, patterns []string) (bool, error) {
-	for _, exclude := range patterns {
-		matched, err := filepath.Match(exclude, relFilePath)
+	matched := false
+
+	for _, pattern := range patterns {
+		pattern = filepath.Clean(pattern)
+		negative := false
+		if empty(pattern) {
+			continue
+		}
+
+		if exclusion(pattern) {
+			if len(pattern) == 1 {
+				logrus.Errorf("Illegal exclusion pattern: %s", pattern)
+				return false, errors.New("Illegal exclusion pattern: !")
+			}
+			negative = true
+			pattern = pattern[1:]
+		}
+
+		match, err := filepath.Match(pattern, relFilePath)
 		if err != nil {
-			logrus.Errorf("Error matching: %s (pattern: %s)", relFilePath, exclude)
+			logrus.Errorf("Error matching: %s (pattern: %s)", relFilePath, pattern)
 			return false, err
 		}
-		if matched {
+		if match {
 			if filepath.Clean(relFilePath) == "." {
```

Comment:
I didn't get a chance to try this, but see if you move this if-statement to the start of this func because technically we should only need to check for this once, and not within the loop.  Hopefully, it doesn't break stuff.  Again, lots of testing needed on this case too because I'm pretty sure we pass "." in the "includes" list, which is what this will deal with.